core/src/main/scala/akka/persistence/r2dbc/internal/BySliceQuery.scala

```diff
@@ -102,20 +191,29 @@ import org.slf4j.Logger
     def nextOffset(state: QueryState, envelope: Envelope): QueryState =
       state.copy(latest = extractOffset(envelope), rowCount = state.rowCount + 1)
 
-    def nextQuery(state: QueryState, toDbTimestamp: Instant): (QueryState, Option[Source[Envelope, NotUsed]]) = {
-      // FIXME why is this rowCount -1 of expected?, see test EventsBySliceSpec "read in chunks"
-      if (state.queryCount == 0L || state.rowCount >= settings.querySettings.bufferSize - 1) {
+    def nextQuery(state: QueryState, endTimestamp: Instant): (QueryState, Option[Source[Envelope, NotUsed]]) = {
+      // Note that we can't know how many events with the same timestamp that are filtered out
+      // so continue until rowCount is 0. That means an extra query at the end to make sure there are no
+      // more to fetch.
+      if (state.queryCount == 0L || state.rowCount > 0) {
```

Comment:
this was an "unrelated" bug: https://github.com/akka/akka-persistence-r2dbc/issues/211